### PR TITLE
doc: rename "recursive search" to "directory scan"

### PR DESCRIPTION
**Pull request by shuber2**
The method 6 of finding the main file is based on a directory scan which is described as "recursive" scan. However, while a recursive scan suggests a (recursive) directory descent, the opposite is true: the unique ascending path of parent directories is scanned.

This is fixed in the following steps:

  1. The method is renamed to "directory scan method"
  2. The method description is adapted to reflect this behavior more clearly. In particular, it is explicitly mentioned that no recursive directory descent is performed.
  3. The option to disable this method is renamed to g:vimtex_disable_directory_scan_main_file_detection

As a side effect, the overall description of the directory scan method is reworked.

As a further side effect, the name of this method is more independent of the algorithmic mechanism, but rather reflects the policy, cf. "policy versus mechanism" paradigm.

### CODE REVIEW

**Comment by shuber2** [comment]
@lervag Sorry, I was not thinking you would be that quick with the work on the other issue. So this pull request will have merge conflict with #2557.

**Comment by lervag** [comment]
> The method 6 of finding the main file is based on a directory scan which is described as "recursive" scan. However, while a recursive scan suggests a (recursive) directory descent, the opposite is true: the unique ascending path of parent directories is scanned.

Interesting point. The reason I use "recursive" here is because it is functionally recursive in the sense that same algorithm is used recursively for complex documents, e.g. `main.tex ← chapter.tex ← section.tex`. If you open `section.tex`, it will being the search from itself, then find that `chapter.tex` contains `section.tex`. It then _recurses_ and finds `main.tex` in the next "iteration". Thus, the point of "recursive" is not about the directory descent at all, it has to do with the project graph of a multi-file LaTeX project.

I'm, curious, though, if you may still have a point with regard to user point of view. What I explained above is an implementation detail, and it is not important (in most cases) for users to care about that.

I'm curious what you think about this.

In any case, I notice that your PR also has good suggestions for how to rephrase and improve the wording, also in general. So I will want to merge it somehow, regardless of the outcome of this discussion!

> 3. The option to disable this method is renamed to g:vimtex_disable_directory_scan_main_file_detection

If we rename an option, we should put the old option in a deprecated list to guide users. That's trivial for me, so I can do it if we go this way. Or I could show you the relevant part in the code.

> As a further side effect, the name of this method is more independent of the algorithmic mechanism, but rather reflects the policy, cf. "policy versus mechanism" paradigm.

Yes, this seems to be relevant to what I wrote above.

> Sorry, I was not thinking you would be that quick with the work on the other issue. So this pull request will have merge conflict with #2558.

I have no problems with merge conflicts. It should be mostly trivial in this case, I think. I'll let you decide if you want to rebase your PR on top of the current master, or if you want me to handle the merge when we've settled the discussions.

File: doc/vimtex.txt [hunk 1]
```diff
@@ -323,13 +323,13 @@ VimTeX supports most multi-file documents and has several methods to locate
 the `main` document. Locating this file is very important, because the main
 file is the one that must be compiled.
 
-The main method for locating the main file uses a recursive search algorithm
-that should find the main LaTeX file in most cases (method 6 in the list
+The main method for locating the main file uses a directory scan algorithm
+that attempts to find the main LaTeX file in most cases (method 6 in the list
```
Comment (lervag) [inline doc/vimtex.txt:327]: I'm not sure I agree with the change "should" to "attempts to". My initial point was to communicate that it will work in most cases. With your proposed change, the wording says it will try to work in most cases, which also means that it will not even try to work in some cases. But that's incorrect.

File: doc/vimtex.txt [hunk 2]
```diff
@@ -447,21 +447,21 @@ Local latexmkrc file specifier~
...
+  If none of the above method gives an appropriate candidate for a main file
+  then a search from the current directory upwards is started. A tex file
+  qualifies as main file if the following requirements are met:
 
-  1. includes the present file (directly or indirectly),
-  2. has a `\documentclass` line near the top, and
-  3. contains `\begin{document}` (or includes a file that does).
+  1. It includes the present file (directly or indirectly).
+  2. It has a `\documentclass` line near the top.
+  3. It contains `\begin{document}` (or includes a file that does).
```
Comment (lervag) [inline doc/vimtex.txt:457]: Thanks; your suggested changes for 453 to 457 are very good!

File: doc/vimtex.txt [hunk 3]
```diff
@@ -447,21 +447,21 @@ Local latexmkrc file specifier~
...
 
-  As long as a project follows these rules, the recursive search should work.
-  Notice, though, that the first point is important and is where the recursive
-  nature is implemented. There must be a "straight" inclusive line from the
-  main file to the present file. If not, then rule 1 is violated.
+  Note that no recursive directory descents are performed to find a main file.
+  That is, if the current file is `./B/chapter.tex` then `./A/main.tex` will not
+  be found as the main file.
```
Comment (lervag) [inline doc/vimtex.txt:461]: I believe my original wording was confusing here. The use of "recursive" is not about directory decending, but about complex projects. I.e., given a project where `main.tex` includes `chapter.tex`, which again includes `section.tex`, if you open `section.tex`, then this is the recursion:

```
find_main(section.tex)
-> find_main(chapter.tex)
  -> find_main(main.tex)
```

where `find_main` will return the file argument if it is a main file, or it will recurse into any file it finds that _includes_ the given file argument.

File: doc/vimtex.txt (hunk 1 above)
Comment (shuber2) [reply doc/vimtex.txt]: I agree, and the recent work on improving the main file detection makes your point even stronger. I have rephrased the paragraph.

File: doc/vimtex.txt (hunk 3 above)
Comment (shuber2) [reply doc/vimtex.txt]: Yes, this is recursive in its nature. That is, the method to expand a tex file recursively through include (and likewise) commands is recursive. This is an implementational aspect of the "directory scan" method. I think what is relevant to the user is the following: The candidate file is expanded and the expanded file needs to meet some requirements (documentclass, begin{document}).

**Comment by shuber2** [comment]
@lervag Thank you for your feedback. I did some work on this:

- Rebase to current master
- Rephrasing to address your feedback
- Adding deprecation information in logic and doc

I also rephased the description of the directory scan method again to make the point on the recursive expansion clearer, including a reference to `g:vimtex_include_indicators`.

When it comes to the overview of methods, the essence of method 6 is that directories are scanned to find a suitable main file. In my opinion, the name of the method should therefore reflect this aspect. When we now describe details when a candidate tex file qualifies as a main file, the aspect of parsing it and recursively expanding it through included tex files becomes relevant. But this is a downstream aspect. And it is an aspect that is true for method 1-5 as well, namely checking whether a candidate tex file is a suitable main file for the current file. What distinguishes method 6 from method 1-5 is that a search for a main file is performed, i.e., directories are scanned for such a file.

**Comment by shuber2** [comment]
A minor rephrasing and a minor spelling fix added.

File: autoload/vimtex/state.vim [hunk 4]
```diff
@@ -237,7 +237,7 @@ function! s:get_main() abort " {{{1
   endif
 
   " Search for main file recursively through include specifiers
-  if !get(g:, 'vimtex_disable_recursive_main_file_detection', 0)
+  if !get(g:, 'vimtex_disable_directory_scan_main_file_detection', 0)
```
Comment (lervag) [inline autoload/vimtex/state.vim:240]: I'm actually considering if we should instead just remove this option alltogether. I don't think _anyone_ uses it. If someone complained, it would be a useful feedback in itself, and I could revert that specific change and use the name you suggest here.

File: autoload/vimtex/state.vim [hunk 5]
```diff
@@ -394,7 +394,7 @@ function! s:get_main_recurse(...) abort " {{{1
   let l:re_filter1 = fnamemodify(l:file, ':t:r')
   let l:re_filter2 = g:vimtex#re#tex_input . '\s*\f*' . l:re_filter1
 
-  " Search through candidates found recursively upwards in the directory tree
+  " Search through candidates found upwards in the directory tree
   let l:results = []
   for l:cand in s:findfiles_recursive('*.tex', fnamemodify(l:file, ':p:h'))
```
Comment (lervag) [inline autoload/vimtex/state.vim:399]: Perhaps we should rename `s:findfiles_recursive` here? And possibly the `s:get_main_recurse` as well? It seems to be in line with your suggestions here.

Comment (lervag) [reply autoload/vimtex/state.vim:399]: Or, perhaps the `s:get_main_recurse` is ok, since it _is_ used recursively. But `s:findfiles_recursive` could perhaps be better named as `s:findfiles_directory_scan`?

File: doc/vimtex.txt [hunk 6]
```diff
@@ -323,23 +323,24 @@ VimTeX supports most multi-file documents and has several methods to locate
...
-main LaTeX file candidate.
+The default method for locating the main file uses a directory scan algorithm
+that searches for a main LaTeX file, see method 6 below. It is expected to
+work in the vast majority of cases.
 
-The methods are tried in the following order:
+There are several alternative methods for specifying the main file that can be
+more flexible and are relevant for certain work flows and use cases. These
+methods all require some explicit declaration of the main file and are
+therefore tried prior the directory scan.
```
Comment (lervag) [inline doc/vimtex.txt:333]: prior _to_ the?

File: doc/vimtex.txt [hunk 7]
```diff
@@ -447,34 +448,38 @@ Local latexmkrc file specifier~
   Note: `@default_files` is a list of files, VimTeX will use the first
         entry that is found.
 
-Recursive search~
-  If no other method provides an appropriate candidate, then the recursive
-  search detects the main LaTeX file by searching for a file in the current
-  and parent directories that
+Directory scan~
+  If none of the above methods gives an appropriate candidate for a main file
```
Comment (lervag) [inline doc/vimtex.txt:452]: * gives -> give
* "for a main file _that directly or indirectly includes_ the present file"

Comment (lervag) [reply doc/vimtex.txt:452]: Or, perhaps simply: "If none of the above methods yield an appropriate main file, then a search ..."?

File: doc/vimtex.txt [hunk 8]
```diff
@@ -447,34 +448,38 @@ Local latexmkrc file specifier~
...
+  of the present file then a search from the current directory upwards is
+  started.
 
-  1. includes the present file (directly or indirectly),
-  2. has a `\documentclass` line near the top of its parsed content, and
-  3. contains `\begin{document}` in its parsed content.
+  To test whether a candidate tex file qualifies as a main file the following
+  happens: First, the candidate tex file is parsed and recursively expanded by
+  all tex files included. Then it is checked whether all of the three
```
Comment (lervag) [inline doc/vimtex.txt:458]: the three "following" requirements

File: doc/vimtex.txt [hunk 9]
```diff
@@ -447,34 +448,38 @@ Local latexmkrc file specifier~
...
-  Notice, though, that the first point is important and is where the recursive
-  nature is implemented. There must be a "straight" inclusive line from the
-  main file to the present file. If not, then rule 1 is violated.
+  1. It includes the present file, either directly or indirectly.
+  2. The expanded content contains a `\documentclass` line near the top.
+  3. The expanded content contains `\begin{document}`.
 
-  Let's consider a common construct that will fail: >
+  Note that the main file itself does not need to contain the `\documentclass`
+  line and `\begin{docment}`; it could instead stem from included tex files.
```
Comment (lervag) [inline doc/vimtex.txt:467]: doc**u**ment

File: doc/vimtex.txt [hunk 10]
```diff
@@ -447,34 +448,38 @@ Local latexmkrc file specifier~
...
 
-    path1/main.tex
-    path2/chapter.tex
-<
-  In this example, the main file does not live in the same folder or a parent
-  folder, and so the recursive search fails.
+  See details on |g:vimtex_include_indicators| for all commands that are
+  considered to include tex files by the parser for the recursive expansion.
```
Comment (lervag) [inline doc/vimtex.txt:470]: Yes, that's good. Thanks!

File: doc/vimtex.txt [hunk 11]
```diff
@@ -447,34 +448,38 @@ Local latexmkrc file specifier~
...
 
-  In these cases where the recursive method fails, one may safely rely on the
-  TeX root directive as an alternative.
+  Note that no recursive directory descents are performed to find a main file.
+  That is, if the current file is `./B/chapter.tex` then `./A/main.tex` will not
+  be found as the main file, because the descent to subdirectory `./A/` is not
+  performed.
+
+  In cases where the directory scan fails, one may instead rely on the TeX
+  root directive as an alternative.
```
Comment (lervag) [inline doc/vimtex.txt:478]: Perhaps we should just refer to _any_ of the methods 1-5, instead of only the tex root alternative?

File: doc/vimtex.txt [hunk 12]
```diff
@@ -447,34 +448,38 @@ Local latexmkrc file specifier~
...
 
   Note: In rare cases, such as if there are _very_ many tex files in the
         directory tree, this method may be slow. One may therefore disable it
-        through the option |g:vimtex_disable_recursive_main_file_detection|.
+        by the option |g:vimtex_disable_directory_scan_main_file_detection|.
```
Comment (lervag) [inline doc/vimtex.txt:482]: I think we can remove this note and the option, as discussed earlier.

File: doc/vimtex.txt [hunk 13]
```diff
@@ -1493,11 +1498,11 @@ OPTIONS                                                        *vimtex-options*
...
-  In rare cases, the recursive method of finding the main file in multi file
-  projects may be slow. This might happen for instance when there are _very_
-  many tex files in the directory tree that is searched. In such cases, one
-  may disable the recursive method by setting this variable to a nonzero
+*g:vimtex_disable_directory_scan_main_file_detection*
+  In rare cases, the directory scan method of finding the main file in multi
+  file projects may be slow. This might happen for instance when there are
+  _very_ many tex files in the directories scanned. In such cases, one may
+  disable the directory scan method by setting this variable to a nonzero
   value.
```
Comment (lervag) [inline doc/vimtex.txt:1506]: See earlier comments.

File: doc/vimtex.txt [hunk 14]
```diff
@@ -6111,6 +6116,11 @@ The following changelog only logs particularly important changes, such as
 changes that break backwards compatibility.  See the git log for the detailed
 changelog.
 
+2022-11-05: Better main file detection algorithm~
+The detection whether a file is a main file has been improved. In course of
+this *g:vimtex_disable_recursive_main_file_detection* has been deprecated in
+favor of |g:vimtex_disable_directory_scan_main_file_detection|.
```
Comment (lervag) [inline doc/vimtex.txt:6122]: If we remove the option, it is still useful to give such a notice.

**Review by lervag — CHANGES_REQUESTED** [review]
Thanks! I think this is going in a good direction that should help make things clearer! I much appreciate your efforts, and I hope you don't find my feedback _too_ nitpicky.

File: autoload/vimtex/state.vim (hunk 5 above)
Comment (shuber2) [reply autoload/vimtex/state.vim:399]: Had the same thoughts. But I think this should be a separate commit to make debugging with git bisect easier later. I will work on that!

File: autoload/vimtex/state.vim (hunk 4 above)
Comment (shuber2) [reply autoload/vimtex/state.vim:240]: One would like to disable it for performance reasons. The feedback on #2558 seems that suggest that this is generally not an issue, but it is hard to estimate for all actual or potential users.
I will implement the removal as a dedicated commit, so it can be cleanly reverted.

File: doc/vimtex.txt (hunk 6 above)
Comment (shuber2) [reply doc/vimtex.txt:333]: yes.

File: doc/vimtex.txt (hunk 7 above)
Comment (shuber2) [reply doc/vimtex.txt:452]: I think "gives" is ok because the indivual, single methods are meant; it can be read as "none of the [individual] above mehtods gives". It would need to be "give" if phrased as "If the above methods give no appropriate…"

File: doc/vimtex.txt (hunk 8 above)
Comment (shuber2) [reply doc/vimtex.txt:458]: yes, makes sense.

File: doc/vimtex.txt [hunk 15]
```diff
@@ -447,34 +448,38 @@ Local latexmkrc file specifier~
...
-  folder, and so the recursive search fails.
+  See details on |g:vimtex_include_indicators| for all commands that are
+  considered to include tex files by the parser for the recursive expansion.
 
-  In these cases where the recursive method fails, one may safely rely on the
-  TeX root directive as an alternative.
+  Note that no recursive directory descents are performed to find a main file.
+  That is, if the current file is `./B/chapter.tex` then `./A/main.tex` will not
+  be found as the main file, because the descent to subdirectory `./A/` is not
+  performed.
```
[the remark this answers is not in the document]
Comment (shuber2) [reply doc/vimtex.txt:475]: Yes, a ascent and descent in the directory _tree_, and (rooted) trees are typically drawn from the root downwards.

File: doc/vimtex.txt (hunk 11 above)
Comment (shuber2) [reply doc/vimtex.txt:478]: That is true, well spotted.

File: doc/vimtex.txt (hunk 12 above)
Comment (shuber2) [reply doc/vimtex.txt:482]: See comment above. I will do that in a separate commit, so it can be atomically reverted-.

File: doc/vimtex.txt (hunk 14 above)
Comment (shuber2) [reply doc/vimtex.txt:6122]: Yes. An the removal commit would rephrase it accordingly.

**Comment by shuber2** [comment]
@lervag All good comments, thanks! And I sympathize with "nitpicking". ;)

File: doc/vimtex.txt (hunk 7 above)
Comment (lervag) [reply doc/vimtex.txt:452]: > I think "gives" is ok because the indivual, single methods are meant; it can be read as "none of the [individual] above mehtods gives". It would need to be "give" if phrased as "If the above methods give no appropriate…".

Ok!

> By the way, it is "If x then y", not "If x, then y". I saw the comma somewhere else arleady in the docs I think.

This is an example of a conditional subordinate conjunction, and I think you need the comma. I found a couple of semi-good references in a quick search:
* https://english.stackexchange.com/questions/373905/how-do-you-punctuate-an-if-then-when-both-the-if-and-the-then-are-implied
* https://www.ego4u.com/en/cram-up/writing/comma?10

File: doc/vimtex.txt (hunk 15 above)
[the remark this answers is not in the document]
Comment (lervag) [reply doc/vimtex.txt:475]: I think that's what confuses me. A tree is _naturally_ pointed upwards, but in these cases they are used _downwards_. 🤷🏻

**Comment by lervag** [comment]
Thanks! I'm glad you don't mind me being pedantic. :)

Looking forward to seeing an updated PR!

**Comment by shuber2** [comment]
@lervag There we go, let's give it another try. I have split the commits up into logical, independent blocks for their own fate. In particular the commit to remove the disable-method-6 option can be easily reverted by itself.

File: doc/vimtex.txt (hunk 15 above)
[the remark this answers is not in the document]
Comment (shuber2) [reply doc/vimtex.txt:475]: By "naturally" you mean biological trees? Yes. In computer science, however, trees are drawn with the root at the top, see https://en.wikipedia.org/wiki/Tree_(data_structure)

Or consider the tree shell command or file browsers.

**Comment by lervag** [comment]
Thanks! I've merged this now after making a few minor changes. Please feel free to let me know if you disagree with them, specifically this commit: https://github.com/lervag/vimtex/commit/af05c7b0c7a4ae05733e1832b39cfa9e57edd694.

**Comment by lervag** [comment]
Oh, and by the way: I very much appreciate this contribution. Documentation is important and hard! And this is actually a very important part of the docs. I hope and believe your contribution here may help make this stuff clearer to both new and existing users in the future, and as such, I think it is a very valuable contribution!

**Comment by shuber2** [comment]
Thank you very much!
